### PR TITLE
lightning: fix "context cancel" overwrites the real error

### LINKED ISSUE #44733: lightning may return "context cancel" error when TiKV disk is slow
## Bug Report

Please answer these questions before submitting your issue. Thanks!

### 1. Minimal reproduce step (Required)

see UT of linked PR

### 2. What did you expect to see? (Required)

error that indicate the TiKV space is not enough

### 3. What did you see instead (Required)

```
[2023/06/16 12:42:17.677 +08:00] [ERROR] [backend.go:353] ["import failed"] [engineTag=`test`.`t`:5] [engineUUID=09383b9a-7e8f-52d1-b6f1-ec86abee1258] [retryCnt=0] [takeTime=668.943958ms] [error="scan regions from start-key:7480000000000000FF5A5F728000000122FF9627480000000000FA, err: rpc error: code = Canceled desc = context canceled: [BR:PD:ErrPDBatchScanRegion]batch scan region"]
```

### PR DESCRIPTION
**Pull request by lance6716**
### What problem does this PR solve?

Issue Number: close #44733

Problem Summary:

### Check List

Tests 

- [x] Unit test
- [ ] Integration test
- [ ] Manual test (add detailed scripts or steps below)
- [ ] No code

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

Documentation

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features
- [ ] Changes MySQL compatibility

### Release note

Please refer to [Release Notes Language Style Guide](https://pingcap.github.io/tidb-dev-guide/contribute-to-tidb/release-notes-style-guide.html) to write a quality release note.

```release-note
None
```

### CODE REVIEW

File: br/pkg/lightning/backend/local/local.go [hunk 1]
```diff
@@ -1145,14 +1145,16 @@ func (local *Backend) generateAndSendJob(
 	for _, jobRange := range jobRanges {
 		r := jobRange
 		eg.Go(func() error {
-			select {
-			case <-egCtx.Done():
+			if egCtx.Err() != nil {
 				return nil
-			default:
 			}
```
Comment (GMHDBJD) [inline br/pkg/lightning/backend/local/local.go:1150]: any difference?

**Review by GMHDBJD — APPROVED** [review]
LGTM

File: br/pkg/lightning/backend/local/local.go [hunk 2]
```diff
@@ -1145,14 +1145,16 @@ func (local *Backend) generateAndSendJob(
...
 
+			failpoint.Inject("beforeGenerateJob", nil)
 			jobs, err := local.generateJobForRange(egCtx, engine, r, regionSplitSize, regionSplitKeys)
 			if err != nil {
+				if common.IsContextCanceledError(err) {
+					return nil
```
Comment (D3Hunter) [inline br/pkg/lightning/backend/local/local.go:1156]: if we return nil here, caller will continue running

File: br/pkg/lightning/backend/local/local.go (hunk 1 above)
Comment (lance6716) [reply br/pkg/lightning/backend/local/local.go:1150]: `select {}` will lock the channel to read the value so performance is better. And code becomes shorter

File: br/pkg/lightning/backend/local/local.go (hunk 2 above)
Comment (lance6716) [reply br/pkg/lightning/backend/local/local.go:1156]: it's expected? because context is `egCtx` and in line 1148 they will fail fast.

**Review by D3Hunter — APPROVED** [review]

File: br/pkg/lightning/backend/local/local.go (hunk 2 above)
Comment (D3Hunter) [reply br/pkg/lightning/backend/local/local.go:1156]: will this happen:
- parent ctx cancelled,`generateAndSendJob` returns nil
- all works returns nil on ctx.Done()
- then we log that engine import success

maybe add ctx.Err into firstErr before return

Comment (lance6716) [reply br/pkg/lightning/backend/local/local.go:1156]: https://github.com/pingcap/tidb/pull/44734/commits/2d3ca6b1585ef9ca2de4770e9fddb3e290f39df4

**Pull request merged**